File: firebase/functions/lib/community/get_community_donations_enabled.dart
import 'dart:async';

import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import '../on_call_function.dart';
import '../utils/firestore_utils.dart';
import '../utils/subscription_plan_util.dart';
import '../utils/utils.dart';
import 'package:data_models/cloud_functions/requests.dart';
import 'package:data_models/firestore/community.dart';
import 'package:data_models/firestore/partner_agreement.dart';

class GetCommunityDonationsEnabled
    extends OnCallMethod<GetCommunityDonationsEnabledRequest> {
  GetCommunityDonationsEnabled()
      : super(
          'GetCommunityDonationsEnabled',
          (jsonMap) => GetCommunityDonationsEnabledRequest.fromJson(jsonMap),
        );

  @override
  Future<Map<String, dynamic>> action(
    GetCommunityDonationsEnabledRequest request,
    CallableContext context,
  ) async {
    return GetCommunityDonationsEnabledResponse(
      donationsEnabled: await _isEnabled(communityId: request.communityId),
    ).toJson();
  }

  Future<bool> _isEnabled({required String communityId}) async {
    final communitySnapshot =
        await firestore.document('/community/$communityId').get();
    orElseNotFound(communitySnapshot.exists);
    final community = Community.fromJson(
      firestoreUtils.fromFirestoreJson(communitySnapshot.data.toMap()),
    );

    if (!kShowStripeFeatures) {
      return false;
    }

    if (!community.settingsMigration.allowDonations) {
      return false;
    }

    final agreementDocs = await firestore
        .collection('partner-agreements')
        .where(
          PartnerAgreement.kFieldCommunityId,
          isEqualTo: communityId,
        )
        .get();
    if (agreementDocs.documents.isNotEmpty) {
      final agreement = PartnerAgreement.fromJson(
        firestoreUtils
            .fromFirestoreJson(agreementDocs.documents.first.data.toMap()),
      );
      return agreement.stripeConnectedAccountActive;
    }

    return false;
  }
}

File: firebase/functions/src/community/get_community_donations_enabled.ts
import * as functions from 'firebase-functions'
import { HttpsError } from 'firebase-functions/lib/providers/https'
import { firestore, firestoreUtils } from '../utils/infra/firestore_utils'
import { OnCallMethod } from '../on_call_function'
import { GetCommunityDonationsEnabledRequest, Community, PartnerAgreement } from '../types'
import { kShowStripeFeatures } from '../utils/subscription_plan_util'
import { orElseNotFound } from '../utils/utils'

export class GetCommunityDonationsEnabled extends OnCallMethod<GetCommunityDonationsEnabledRequest> {
    constructor() {
        super('GetCommunityDonationsEnabled', (json) => json as GetCommunityDonationsEnabledRequest)
    }

    async action(
        request: GetCommunityDonationsEnabledRequest,
        _context: functions.https.CallableContext
    ): Promise<{ donationsEnabled: boolean }> {
        return { donationsEnabled: await this._isEnabled({ communityId: request.communityId }) }
    }

    private async _isEnabled({ communityId }: { communityId: string }): Promise<boolean> {
        const communitySnapshot = await firestore.doc(`/community/${communityId}`).get()
        orElseNotFound(communitySnapshot.exists)
        const community = communitySnapshot.data() as Community

        if (!kShowStripeFeatures) return false

        const communitySettings = community.communitySettings
        if (!communitySettings?.allowDonations) return false

        const agreementDocs = await firestore
            .collection('partner-agreements')
            .where('communityId', '==', communityId)
            .get()

        if (!agreementDocs.empty) {
            const agreement = agreementDocs.docs[0].data() as PartnerAgreement
            return agreement.stripeConnectedAccountActive ?? false
        }

        return false
    }
}
